### ISSUE TITLE
Scene RigidObjects get overwritten to ProjectileConfig.hide_z  one sim step after reset (IsaacLab backend)

### ISSUE THREAD

**Issue by jbae49** [NVlabs/ProtoMotions#210]
### Problem
On the IsaacLab backend, a task with scene-based `RigidObject` instances (we use `BoxSceneObject` via `scene_lib`, one box per env, ~4096 boxes on a 10 m-spacing grid, alongside a single G1 robot articulation per env) loses its scene object state mid-training. The failure is specific to **scene objects**; the robot articulation and projectile objects are unaffected.

One physics step after a reset (no intervening reset or termination), every scene box's root pose snaps to exactly `(0, 0, -2.0)`, which is`ProjectileConfig.hide_z`. Manifests as `obj_pos_err` exploding to the mean distance between each env's scene-grid position and the world origin (~1282 m averaged over a 4096-env grid of 10 m spacing; this is just `mean(norm(scene_grid_pos))` across envs, not anything physical). All envs terminate simultaneously, `episode_length` collapses to 1.

### Bisect
Same seed, same config, same data:
| `ProjectileConfig.num_projectiles` | Outcome |
|---|---|
| 5 (default) | fails at epoch 47–107 |
| 0 | clean through epoch 946+, `obj_pos_err ≈ 6 cm` |

Projectile code path is definitely involved, but we couldn't isolate the exact aliasing between projectile writes and scene-object state.

### Diagnostic 
We added rate-limited hooks around `isaaclab/simulator.py::_set_simulator_env_state`: **At the moment of reset write, write to the scene box is clean** (across all 4096 envs):

|intended_xyz - self._object[0].data.root_pos_w| = 0 
|intended_xyz - self._object[0].root_physx_view pose| = 0 
|data.root_pos_w - root_physx_view pose| = 0

**One env step later, same boxes (`progress_buf == 1`, no reset, no termination)**:
env=4093 cur_buf=(0, 0, -2.00) physx=(0, 0, -2.00) ref_world=(180.27, 2290.16, 0.17) → err=2297 m

Both CPU buffer and PhysX direct read report `(0, 0, -2.0)`. Not a buffer-vs-PhysX desync, not gravity (2 m vertical + ~2 km horizontal in one env step is impossible). The reset write is correct; something overwrites it during the subsequent simulation step.

### What we patched locally (workaround)
We force `num_projectiles = 0` via env var. This avoids the issue but **disables projectile disturbances entirely**, which are required for our intended training setup (external perturbation robustness). It also needs a second patch because  `base_simulator/simulator.py::_init_projectiles` unconditionally does `self._proj_config = ProjectileConfig()`, overwriting any customization done in the backend `__init__`.

### What we'd like to fix (but couldn't)
The actual aliasing path in `isaaclab/simulator.py` that lets projectile code reach scene-object state. 

### Questions 
1. Are scene `RigidObject` views (`self._object[i]`) and projectile `RigidObject` views (`self._projectile_objects[i]`) fully isolated  in this backend; separate prim paths, separate PhysX views, non-overlapping `env_ids` masks?
2. Could `scene.write_data_to_sim()` flush any projectile-pool default state (initialized at `hide_z = -2.0`) into scene-object state between our reset write and the next `get_transforms()` read? I.e. is there any shared default-root-state buffer or fabric path between the two object families?

**Comment by tesslerc** [comment]
We were unable to reproduce the bug you reported locally (re-shifting objects).
If you could share a reproducable experiment-file/script we can verify on our end and help find a solution.

---

https://github.com/NVlabs/ProtoMotions/pull/211
This should be a partial fix, could you take a look?

This fixes the overwrite bug on the projectile config between base_simulator and the different sim implementations.
It also allows configurable projectile logic parameters.

Regarding the questions:
1. Yes the views are fully isolated. You can see [here](https://github.com/NVlabs/ProtoMotions/blob/b59d44f5b3be5a9a33a3fac6a82d40e5aecaba63/protomotions/simulator/isaaclab/simulator.py#L148) we store the objects and [here](https://github.com/NVlabs/ProtoMotions/blob/b59d44f5b3be5a9a33a3fac6a82d40e5aecaba63/protomotions/simulator/isaaclab/simulator.py#L158) the projectiles. Similar logic in the IsaacGym implementation.
2. Should not hit the wrong assets. To the best of our knowledge, `write_root_state_to_sim` updates the states and then `write_data_to_sim` pushes the update to sim. These buffers are then cleared and there is no default override that happens unintentionally.

**Comment by jbae49** [comment]
Thanks for the partial fix and the detailed explanation, really appreciate it. I've confirmed that setting `simulator.projectile.num_projectiles = 0` (or our local PROTOMOTIONS_DISABLE_PROJECTILES=1 env-var workaround) makes the 1282 m blow-up disappear entirely on our side, so the new override is definitely useful. 

That said, with projectiles enabled we can still trigger the corruption reliably, and the scene-object root buffer ends up at z = -2.00 m (byte-for-byte equal to ProjectileConfig.hide_z), with PhysX agreeing to machine precision. So while the view aliasing may not be the culprit, something is still overwriting the scene-object buffer to the projectile hide pose once training hits a certain state.

To help narrow it down, I've put together a minimal reproduction bundle with the exact motion/scene .pt files, our experiment script, annotated FDUMP logs from the failing epoch, a wandb export script, and visualisation scripts:

Bundle (1.3 MB, attached to this comment) — [repro_1282m_bug_core.tar.gz](https://github.com/user-attachments/files/26951048/repro_1282m_bug_core.tar.gz)

Checkpoint from the crashing run (45 MB, Google Drive) — https://drive.google.com/file/d/1kUNZg_fuFJrJ7bDAFHA5Y8hvhgh_Ib-P/view?usp=sharing
(drop last.ckpt into repro_1282m_bug/checkpoint/ to enable the fast repro path)

The bundle's README.md covers runtime environment, exact training command, step-by-step repro (Path A a couple mins with the checkpoint, Path B full train from scratch -> took under 2 hours on a single RTX 4090 before the first blow-up around epoch 144.), the FDUMP evidence

(Minor correction: the `Runtime environment` section in the bundled `README.md` says "RTX 4070 12 GB" — that's a typo, the actual hardware I used was a single RTX 4090 (24 GB VRAM))

Happy to add more instrumentation or test a branch if it helps narrow this down. Thanks again!

**Comment by tesslerc** [comment]
Thanks. I think I'm missing `protomotions.envs.box_lift_env_v2.BoxLiftEnvV2`. Potentially other env changes that might be needed?

In the meantime, if you train with much less envs, does the same issue happen?
Do you see any prints complaining about lost contacts or physx buffer overflows? (IsaacLab prints)

**Comment by jbae49** [comment]
Thanks! Attaching [box_lift_env_v2.py](https://github.com/user-attachments/files/27000466/box_lift_env_v2.py); sorry for the miss. 

On your two follow-up questions:

- Fewer envs: I haven't tested with a smaller env count yet. Happy to try 512 / 1024 envs tomorrow if it'd help narrow this down.
- PhysX / IsaacLab warnings: I re-grepped the crashing run's full stdout (lost contact, contact buffer, overflow, dropped, truncate, any physx.*warn) and found nothing beyond the usual startup warnings (GLFW / CUDA_VISIBLE_DEVICES / gpu.foundation). No contact-buffer or GPU-report-flag overflow messages anywhere in the epochs leading up to the crash.

For what it's worth, one thing that keeps us suspicious of a structural overwrite rather than a buffer-capacity issue: the scene-object root lands at exactly (0, 0, -2.00) (bit-exact to ProjectileConfig.hide_z), with PhysX agreeing to machine precision. That's what hinted at the projectile path rather than a contact-buffer truncation.

**Comment by tesslerc** [comment]
Thanks. But now also missing `from protomotions.envs.box_lift_env import BoxLiftEnv`.
I hacked around this, replacing `BoxLiftEnv` with `BaseEnv` and tried to reproduce.

I ran with the given checkpoint for 5000 epochs. 4096 envs with a forced reset every 200 epochs. Using a single RTX 5090.
The bug did not reproduce.

I'll try again with the proper `BoxLiftEnv` to see if that changes the results.

**Comment by jbae49** [comment]
Apologies for the miss. Attaching [box_lift_env.py](https://github.com/user-attachments/files/27033111/box_lift_env.py) !

**Comment by tesslerc** [comment]
Thanks!

We have now reproduced it on our end. PR fixing it is here: https://github.com/NVlabs/ProtoMotions/pull/214
Could you please test on your end and see if it indeed fixes your issue?

TL;DR -- all projectiles are respawned at (0,0,-2) which leads to thousands of overlapping objects when all envs reset together. This seems to lead to memory corruption.
